### PR TITLE
[IMP] data_analysis: detection of headers

### PR DESCRIPTION
**Pull request by matho-odoo**
## Description:

description of this task, what is implemented and why it is implemented that way.

Task: [6516198](https://www.odoo.com/odoo/2328/tasks/6516198)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/helpers/figures/charts/chart_suggestion_engine.ts [hunk 1]
```diff
@@ -150,23 +150,29 @@ function interestingCellsXc(col: ColumnAnalysis): {
   return { firstCellXC, lastCellXC, prevCellXC };
 }
 
-function getCategoryHeader(getters: Getters, catCol: ColumnAnalysis): string {
-  let catHeader = catCol.header ?? "";
-  if (!catHeader && catCol.rowCount > 0) {
+function getCategoryHeader(getters: Getters, catCol: ColumnAnalysis, hasTitle: boolean): string {
```
Comment (anhe-odoo) [inline src/helpers/figures/charts/chart_suggestion_engine.ts:153]: Now we always have a title, while before your changes we sometimes had "" as title, leading to no title for the scorecard, which was intended

File: src/helpers/figures/charts/chart_suggestion_engine.ts [hunk 2]
```diff
@@ -150,23 +150,29 @@ function interestingCellsXc(col: ColumnAnalysis): {
...
-  return catHeader;
+  return _t("Category");
 }
 
 /** Pattern A — Single numeric column */
 function buildSingleNumberContext([col]: ColumnAnalysis[], getters: Getters): SingleNumberContext {
   const title = col.header ?? "";
   const { firstCellXC, lastCellXC, prevCellXC } = interestingCellsXc(col);
-  const source = rangeSource([dataset(col.zone, getters)], col.hasHeader);
+  const source = rangeSource([dataset(col.zone, getters)], col.headerInZone);
```
Comment (anhe-odoo) [inline src/helpers/figures/charts/chart_suggestion_engine.ts:175]: Not sure of the change of this field name :/

File: src/helpers/figures/charts/chart_suggestion_engine.ts [hunk 3]
```diff
@@ -283,8 +289,8 @@ function buildCategoricalVsPercentageContext(
   getters: Getters
 ): CategoricalVsPercentageContext {
   const labelRange = getUnboundRange(getters, catCol.zone);
-  const hasTitle = pctCol.hasHeader;
-  const categoryHeader = hasTitle ? getCategoryHeader(getters, catCol) : _t("Category");
+  const hasTitle = pctCol.headerInZone;
+  const categoryHeader = getCategoryHeader(getters, catCol, hasTitle);
```
Comment (anhe-odoo) [inline src/helpers/figures/charts/chart_suggestion_engine.ts:293]: The advantage of the ternary was that sometimes we wanted to have "Category" and sometimes we wanted an empty title, "au cas par cas"

File: src/helpers/data_analysis.ts [hunk 4]
```diff
@@ -51,11 +52,27 @@ function analyzeColumn(zone: Zone, getters: Getters): ColumnAnalysis {
...
-  // Header: first cell is text AND rest has at least one non-text, non-empty cell
-  const hasHeader =
-    firstCell.type === CellValueType.text && rest.some((c) => c.type !== CellValueType.text);
+  let header: string | undefined;
+  let dataCells: EvaluatedCell[] = [];
+  let headerInZone = false;
+  const tableHeaderPosition = getColumnTableHeaderPosition(sheetId, zone, getters);
 
-  const dataCells: EvaluatedCell[] = hasHeader ? rest : cells;
+  if (tableHeaderPosition) {
```
Comment (anhe-odoo) [inline src/helpers/data_analysis.ts:60]: This part and the `getColumnTableHeaderPosition` function are probably the only useful parts of this PR :) I think we can keep these and revert all the other changes

File: src/helpers/table_helpers.ts [hunk 5]
```diff
@@ -324,3 +327,35 @@ function getTableElementZones(
...
+    return undefined;
+  }
+  const tables = getters.getTablesOverlappingZones(sheetId, [zone]);
+  if (tables.length !== 1) {
+    return undefined;
+  }
+  const table = tables[0];
+  const firstCell = { sheetId, col: zone.left, row: zone.top };
+  const lastCell = { sheetId, col: zone.left, row: zone.bottom };
+  if (firstCell.row < table.range.zone.top || lastCell.row > table.range.zone.bottom) {
```
Comment (anhe-odoo) [inline src/helpers/table_helpers.ts:346]: Shouldn't we use the header only in the strict case where the first cell in zone is exactly the first of the tale col ?

File: src/helpers/figures/charts/chart_suggestion_engine.ts (hunk 1 above)
Comment (matho-odoo) [reply src/helpers/figures/charts/chart_suggestion_engine.ts]: Only when we call this function which is not all the time

File: src/helpers/figures/charts/chart_suggestion_engine.ts (hunk 3 above)
Comment (matho-odoo) [reply src/helpers/figures/charts/chart_suggestion_engine.ts]: I don't see wich case would be wrong

File: src/helpers/figures/charts/chart_suggestion_engine.ts (hunk 1 above)
Comment (anhe-odoo) [reply src/helpers/figures/charts/chart_suggestion_engine.ts]: Tu as raison

File: src/helpers/table_helpers.ts [hunk 6]
```diff
@@ -324,3 +327,35 @@ function getTableElementZones(
 
   return zones;
 }
+
+export function getColumnTableHeaderPosition(
```
Comment (hokolomopo) [inline src/helpers/table_helpers.ts]: Nitpick: Mayble a smll docstring explaining what it does (zone need to be inside a single table & all non empty cells should be in table).

I'd also move the helper to dataAnalysis (and not export it), the behavior is very specific to the data analysis feature, not really a generic table behaviour.

File: src/helpers/data_analysis.ts [hunk 7]
```diff
@@ -51,11 +52,27 @@ function analyzeColumn(zone: Zone, getters: Getters): ColumnAnalysis {
...
+    header = getters.getCellText(tableHeaderPosition) || undefined;
+    dataCells = cells.filter((c) => c.position && c.position.row > tableHeaderPosition.row);
+    headerInZone =
+      !!header && tableHeaderPosition.row >= zone.top && tableHeaderPosition.row <= zone.bottom;
```
Comment (hokolomopo) [inline src/helpers/data_analysis.ts]: I think it's a bit wrong. We use headerInZone to determine whether the chart dataset have a header row or not. But for tables with multiple header rows, the first x rows might need to be ignored. Or even with a single table header row, we might select empty cells above the table.

At the same time taking the sub-header as a title is a bit strange if we select the whole table column as the chart dataset.  

We probably want to change the zone in those case, to only take data starting at the table sub-header and ending at the bottom of the table.

Image

File: src/helpers/figures/charts/chart_suggestion_engine.ts [hunk 8]
```diff
@@ -175,7 +181,7 @@ function buildSinglePercentageContext(
   [col]: ColumnAnalysis[],
   getters: Getters
 ): SinglePercentageContext {
-  const hasTitle = col.hasHeader;
+  const hasTitle = col.headerInZone;
```
Comment (hokolomopo) [inline src/helpers/figures/charts/chart_suggestion_engine.ts]: In the ColumnAnalysis I'd change the name to `dataHasHeader(row)` to be clearer. 

Also in all those helper, the variable `hasTitle` is confusing; it having nothing to do with the actual chart title. I'd also rename it to have something to do with the datasets.

File: tests/helpers/data_analysis_helpers.test.ts [hunk 9]
```diff
@@ -65,19 +70,122 @@ describe("analyzeColumns", () => {
...
+    const model = createModelFromGrid({ A1: "100", A2: "200", A3: "300" });
+    createTable(model, "A1:A3", { numberOfHeaders: 0 });
+    const col = analyzeColumn(model, "A1:A3");
+    expect(col.headerInZone).toBe(false);
+    expect(col.header).toBeUndefined();
+    expect(col.rowCount).toBe(3);
+  });
+
+  test("the analyzed zone extends past the table", () => {
+    const model = createModelFromGrid({
```
Comment (hokolomopo) [inline tests/helpers/data_analysis_helpers.test.ts]: nitpick: improve the test names to say what should happen when the table has no header/the zone extend past the bale.

Also maybe a test that we can expand past the table if the rows are all empty

**Review by hokolomopo — COMMENTED** [review]
👋

**Comment by hokolomopo** [comment]
@matho-odoo small rebase then we can r+ :)

**Comment by hokolomopo** [comment]
@robodoo r+
